HOTT-1483 Show ancestors tree on Subheading page

jebw opened the pull request:
### Jira link

[HOTT-1483](https://transformuk.atlassian.net/browse/HOTT-1483)

### What?

I have added/removed/altered:

- [x] Moved ancestors partial into common goods_nomenclatures folder to show it applies to more then declarables/commodities
- [x] Show the ancestors tree on the Subheading page
- [x] Changed the Subheading tree to only show commodities below the Subheading being viewed
- [x] Default Commodity Tree on Heading and Subheading pages to show open instead of closed

### Why?

I am doing this because:

- It will aid users with navigating the site

### Have you? (optional)

- [x] Reviewed view changes with stake holders
- [x] Validated mobile responsive behaviour of view changes

jebw committed:
HOTT-1483 Show ancestors tree on Subheading page

Moved ancestors partial into common goods_nomenclatures folder to show
it applies to more then declarables/commodities

alexdesi commented on spec/factories/subheading_factory.rb:
```diff
@@ -1,13 +1,28 @@
     section { attributes_for(:section) }
-    chapter { attributes_for(:chapter) }
-    commodities { [] }
+
+    chapter do
+      attributes_for :chapter,
+                     goods_nomenclature_item_id: "#{goods_nomenclature_item_id.first(2)}00000000"
+    end
+
+    heading do
+      attributes_for :heading,
+                     goods_nomenclature_item_id: "#{goods_nomenclature_item_id.first(4)}000000"
+    end
```
Should these be in traits to not slow down the other tests that don't need those?

jebw replied:
This is used to generated models simulating what we get back from the API, the API would include the heading data so the absence of heading makes the generated subheading not particularly functional or reflective of a 'real' API response. 

I don't think it should impact the performance in any measurable way - its just generating a hash of test data, not persisting anything.

alexdesi reviewed (dismissed):
it looks good to me ✅ 
only a comment

jebw committed:
HOTT-1483 Added #parent to commodites

When downloaded as part of a tree on a heading or subheading

jebw committed:
HOTT-1483 Show all ancestors on subheading page

jebw committed:
HOTT-1483 Avoid global state in commodities tree partial

jebw committed:
HOTT-1483 Only show commodities below Subheading

on Subheadings page

willfish commented on app/models/subheading.rb:
```diff
@@ -41,6 +41,22 @@ def to_s
     formatted_description || description
   end
 
+  def ancestors
```
We should just return this from the backend

willfish commented on spec/factories/heading_factory.rb:
```diff
@@ -1,12 +1,24 @@
 FactoryBot.define do
   factory :heading do
+    transient do
+      commodity_count { nil }
```
```suggestion
      commodity_count { 0 }
```

willfish commented on spec/factories/heading_factory.rb:
```diff
@@ -1,12 +1,24 @@
     goods_nomenclature_item_id { '0101000000' }
 
-    commodities { [] }
+    commodities { attributes_for_list :commodity, commodity_count || 0 }
```
```suggestion
    commodities { attributes_for_list :commodity, commodity_count }
```

jebw committed:
HOTT-1483 Default commodity trees to be open

Previously when the user had not explicitly chosen with open or closed
they defaulted to closed

willfish commented on spec/factories/heading_factory.rb:
```diff
@@ -1,12 +1,24 @@
+
+    trait :with_commodity_tree do
+      commodities do
+        attributes_for_list(:commodity, commodity_count || 2) do |c, i|
```
I'm assuming c is a commodity and i is the index of the list of the commodities starting out at 0?

jebw replied:
I'll change them to commodity and index

willfish commented on spec/models/subheading_spec.rb:
```diff
@@ -86,9 +86,26 @@
+
+    let :commodities do
+      attributes_for_list :commodity, 5 do |c, i|
+        c[:parent_sid] = (i > 0 ? c[:goods_nomenclature_sid] - 1 : nil)
```
There's a lot of baked in programmatic understanding in the fixtures, here. Is there any chance we could just be explicit with some sids to simulate real data/also make the fixtures a bit more dumb?

jebw replied:
Problem with hard coding sid's (or any ids) is that you can only use the factory once per spec (you couldn't create 2 subheadings with tree's of commodities under them for example).

jebw replied:
I'm updating this to use manually specified ids but since I'm having to specify more ids I'm pretty sceptical that it _reduces_ the knowledge subheading has of commodity tree's - it seems to result in an increase

jebw committed:
HOTT-1483 Use backend api for ancestors list
